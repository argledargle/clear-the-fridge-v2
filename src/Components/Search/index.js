import React, { useState, useReducer } from 'react'
import _pull from 'lodash/pull'

const url = 'https://api.edamam.com/search?'
const appId = process.env.REACT_APP_APPID
const appKey = process.env.REACT_APP_API_KEY

const dietArray = [
  {
    name: 'Vegan',
    value: 'vegan'
  },
  {
    name: 'Vegetarian',
    value: 'vegetarian'
  },
  {
    name: 'Sugar Conscious',
    value: 'sugar-conscious'
  },
  {
    name: 'Peanut Free',
    value: 'peanut-free'
  },
  {
    name: 'Tree-nut Free',
    value: 'tree-nut-free'
  },
  {
    name: 'Alcohol Free',
    value: 'alcohol-free'
  }
]

const Search = ({ callback }) => {
  const [input, setInput] = useState(undefined)

  const handleInputChange = e =>
    setInput({
      ...input,
      [e.currentTarget.name]: e.currentTarget.value
    })

  const [checkedDietArray, dispatch] = useReducer(
    (checkedDietArray, { type, value }) => {
      switch (type) {
        case 'add':
          return [...checkedDietArray, value]
        case 'remove':
          return _pull(checkedDietArray, value)
        default:
          return checkedDietArray
      }
    },
    []
  )

  function buildSearchQuery (e) {
    e.preventDefault()
    if (
      input === undefined ||
      input.textInput === undefined ||
      input.textInput === ''
    ) {
      // eslint-disable-next-line no-undef
      alert('You need to enter something into the text field.')
      return null
    }
    const searchUrl = formatQueryParams(input)
    getResults(searchUrl)

    return formatQueryParams(input)
  }

  function getResults (url) {
    // eslint-disable-next-line no-undef
    fetch(url)
      .then(response => response.json())
      .then(data => callback(data))
  }

  function formatQueryParams (params) {
    const healthQuery = checkedDietArray
      .map(item => {
        return `health=${item}&`
      })
      .toString()
    const queryStringComponents = {
      app_key: appKey,
      app_id: appId
    }

    if (params.textInput !== '' || params.textInput !== undefined) {
      queryStringComponents.q = params.textInput
    }

    if (params.dietInput !== '' && params.dietInput !== undefined) {
      queryStringComponents.diet = params.dietInput
    }

    const queryString = Object.keys(queryStringComponents)
      .map(key => `${key}=${queryStringComponents[key]}`)
      .join('&')
    const newUrl = url + healthQuery + queryString
    return newUrl
  }

  return (
    <div className='max-w-lg flex p-6 bg-gray-100 mt-10 rounded-lg shadow-md mx-auto'>
      <div className='pt-1'>
        <h1 className='text-xl text-blue-700 leading-tight'>
          What do you want to use up?
        </h1>
        <p className='text-base text-gray-700 leading-normal'>
          Type in the ingredients you're looking to clear out from your kitchen,
          select a diet and click search.
        </p>
        <form className='mx-auto'>
          <div className='flex flex-row justify-between'>
            <input
              className='rounded-sm border border-gray-600'
              name='textInput'
              type='text'
              placeholder='Chicken, rice'
              id='search-term'
              required
              onChange={handleInputChange}
            />
            <select
              className='rounded-sm border-gray-600 border'
              name='dietInput'
              onChange={handleInputChange}
            >
              <option value='' defaultValue>
                Diet
              </option>
              <option value='balanced'>Balanced</option>
              <option value='high-protein'>High-Protein</option>
              <option value='low-carb'>Low-Carb</option>
              <option value='low-fat'>Low-Fat</option>
            </select>
            <input
              type='submit'
              value='Search'
              className='rounded-sm border-gray-600 border'
              onClick={buildSearchQuery}
            />
          </div>
          <fieldset>
            <legend className='text-lg'>Optional health choices:</legend>
            <div className='grid grid-flow-col grid-cols-2 grid-rows-3'>
              {dietArray.map(item => {
                return (
                  <div className='inline' key={item.value}>
                    <input
                      className='mx-1'
                      type='checkbox'
                      name={item.value}
                      value={item.value}
                      id={item.value}
                      onChange={() =>
                        dispatch({
                          type: checkedDietArray.includes(item.value)
                            ? 'remove'
                            : 'add',
                          value: item.value
                        })}
                    />
                    <label
                      className='text-base text-gray-700 leading-normal'
                      htmlFor={item.value}
                    >
                      {item.name}
                    </label>
                  </div>
                )
              })}
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  )
}

export default Search
