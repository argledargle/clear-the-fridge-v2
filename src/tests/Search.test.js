import React from 'react'
import { render, screen } from '@testing-library/react'
import Search from '../Components/Search'
import userEvent from '@testing-library/user-event'

describe('Search', () => {
  it('clicking on the checkboxes should fire off dispatch function', async () => {
    render(<Search />)
    const checkbox = screen.getByTestId('vegetarian')
    userEvent.click(checkbox)
    expect(checkbox).toBeChecked()
  })

  it('Runs a full search by changing options multiple times', async () => {
    const newScreen = render(<Search />)
    const vegetarianCheckbox = newScreen.getByTestId('vegetarian')
    const veganCheckbox = newScreen.getByTestId('vegan')
    const treeNutCheckbox = newScreen.getByTestId('tree-nut-free')
    const highProtein = newScreen.getByTestId('highProtein')
    const input = newScreen.getByPlaceholderText('Chicken, rice')
    const submit = newScreen.getByTestId('submit')

    userEvent.click(vegetarianCheckbox)
    userEvent.click(veganCheckbox)
    expect(vegetarianCheckbox).toBeChecked()
    expect(veganCheckbox).toBeChecked()
    userEvent.click(veganCheckbox)
    expect(veganCheckbox).not.toBeChecked()

    userEvent.click(highProtein)
    userEvent.click(treeNutCheckbox)

    userEvent.click(input)
    userEvent.type(input, 'chicken, cheese, rice')
    userEvent.click(submit)

    expect(newScreen).toMatchSnapshot()
  })
})
