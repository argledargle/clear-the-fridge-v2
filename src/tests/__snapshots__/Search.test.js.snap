// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Runs a full search by changing options multiple times 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="max-w-lg flex p-6 bg-gray-100 mt-10 rounded-lg shadow-md mx-auto"
      >
        <div
          class="pt-1"
        >
          <h1
            class="text-xl text-blue-700 leading-tight"
          >
            What do you want to use up?
          </h1>
          <p
            class="text-base text-gray-700 leading-normal"
          >
            Type in the ingredients you're looking to clear out from your kitchen, select a diet and click search.
          </p>
          <form
            class="mx-auto"
          >
            <div
              class="flex flex-row justify-between"
            >
              <input
                class="rounded-sm border border-gray-600"
                id="search-term"
                name="textInput"
                placeholder="Chicken, rice"
                required=""
                type="text"
              />
              <select
                class="rounded-sm border-gray-600 border"
                data-testid="dietSelect"
                name="dietInput"
              >
                <option
                  value=""
                >
                  Diet
                </option>
                <option
                  value="balanced"
                >
                  Balanced
                </option>
                <option
                  data-testid="highProtein"
                  value="high-protein"
                >
                  High-Protein
                </option>
                <option
                  data-testid="lowCarb"
                  value="low-carb"
                >
                  Low-Carb
                </option>
                <option
                  value="low-fat"
                >
                  Low-Fat
                </option>
              </select>
              <input
                class="rounded-sm border-gray-600 border"
                data-testid="submit"
                type="submit"
                value="Search"
              />
            </div>
            <fieldset>
              <legend
                class="text-lg"
              >
                Optional health choices:
              </legend>
              <div
                class="grid grid-flow-col grid-cols-2 grid-rows-3"
              >
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="vegan"
                    id="vegan"
                    name="vegan"
                    type="checkbox"
                    value="vegan"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="vegan"
                  >
                    Vegan
                  </label>
                </div>
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="vegetarian"
                    id="vegetarian"
                    name="vegetarian"
                    type="checkbox"
                    value="vegetarian"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="vegetarian"
                  >
                    Vegetarian
                  </label>
                </div>
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="sugar-conscious"
                    id="sugar-conscious"
                    name="sugar-conscious"
                    type="checkbox"
                    value="sugar-conscious"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="sugar-conscious"
                  >
                    Sugar Conscious
                  </label>
                </div>
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="peanut-free"
                    id="peanut-free"
                    name="peanut-free"
                    type="checkbox"
                    value="peanut-free"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="peanut-free"
                  >
                    Peanut Free
                  </label>
                </div>
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="tree-nut-free"
                    id="tree-nut-free"
                    name="tree-nut-free"
                    type="checkbox"
                    value="tree-nut-free"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="tree-nut-free"
                  >
                    Tree-nut Free
                  </label>
                </div>
                <div
                  class="inline"
                >
                  <input
                    class="mx-1"
                    data-testid="alcohol-free"
                    id="alcohol-free"
                    name="alcohol-free"
                    type="checkbox"
                    value="alcohol-free"
                  />
                  <label
                    class="text-base text-gray-700 leading-normal"
                    for="alcohol-free"
                  >
                    Alcohol Free
                  </label>
                </div>
              </div>
            </fieldset>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="max-w-lg flex p-6 bg-gray-100 mt-10 rounded-lg shadow-md mx-auto"
    >
      <div
        class="pt-1"
      >
        <h1
          class="text-xl text-blue-700 leading-tight"
        >
          What do you want to use up?
        </h1>
        <p
          class="text-base text-gray-700 leading-normal"
        >
          Type in the ingredients you're looking to clear out from your kitchen, select a diet and click search.
        </p>
        <form
          class="mx-auto"
        >
          <div
            class="flex flex-row justify-between"
          >
            <input
              class="rounded-sm border border-gray-600"
              id="search-term"
              name="textInput"
              placeholder="Chicken, rice"
              required=""
              type="text"
            />
            <select
              class="rounded-sm border-gray-600 border"
              data-testid="dietSelect"
              name="dietInput"
            >
              <option
                value=""
              >
                Diet
              </option>
              <option
                value="balanced"
              >
                Balanced
              </option>
              <option
                data-testid="highProtein"
                value="high-protein"
              >
                High-Protein
              </option>
              <option
                data-testid="lowCarb"
                value="low-carb"
              >
                Low-Carb
              </option>
              <option
                value="low-fat"
              >
                Low-Fat
              </option>
            </select>
            <input
              class="rounded-sm border-gray-600 border"
              data-testid="submit"
              type="submit"
              value="Search"
            />
          </div>
          <fieldset>
            <legend
              class="text-lg"
            >
              Optional health choices:
            </legend>
            <div
              class="grid grid-flow-col grid-cols-2 grid-rows-3"
            >
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="vegan"
                  id="vegan"
                  name="vegan"
                  type="checkbox"
                  value="vegan"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="vegan"
                >
                  Vegan
                </label>
              </div>
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="vegetarian"
                  id="vegetarian"
                  name="vegetarian"
                  type="checkbox"
                  value="vegetarian"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="vegetarian"
                >
                  Vegetarian
                </label>
              </div>
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="sugar-conscious"
                  id="sugar-conscious"
                  name="sugar-conscious"
                  type="checkbox"
                  value="sugar-conscious"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="sugar-conscious"
                >
                  Sugar Conscious
                </label>
              </div>
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="peanut-free"
                  id="peanut-free"
                  name="peanut-free"
                  type="checkbox"
                  value="peanut-free"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="peanut-free"
                >
                  Peanut Free
                </label>
              </div>
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="tree-nut-free"
                  id="tree-nut-free"
                  name="tree-nut-free"
                  type="checkbox"
                  value="tree-nut-free"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="tree-nut-free"
                >
                  Tree-nut Free
                </label>
              </div>
              <div
                class="inline"
              >
                <input
                  class="mx-1"
                  data-testid="alcohol-free"
                  id="alcohol-free"
                  name="alcohol-free"
                  type="checkbox"
                  value="alcohol-free"
                />
                <label
                  class="text-base text-gray-700 leading-normal"
                  for="alcohol-free"
                >
                  Alcohol Free
                </label>
              </div>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
